src/core/common/binary_search.hpp

```diff
@@ -0,0 +1,127 @@
+/*
+ *  Copyright (c) 2022, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *  This file provides a generic binary search and related helper functions.
+ */
+
+#ifndef BINARY_SEARCH_HPP_
+#define BINARY_SEARCH_HPP_
+
+#include "openthread-core-config.h"
+
+#include <stdint.h>
+
+namespace ot {
+
+class BinarySearch
+{
+public:
+    /**
+     * This template method performs binary search in a given sorted table array to find an entry matching a given key.
+     *
+     * @note This method requires the array to be sorted, otherwise its behavior is undefined.
+     *
+     * @tparam Key         The type of `Key` to search for.
+     * @tparam Entry       The table `Entry` type.
+     * @tparam kLength     The array length (number of entries in the array).
+     *
+     * The `Entry` class MUST provide the following method to compare the entry against a given key.
+     *
+     *    int Entry::Compare(const Key &aKey) const;
+     *
+     * The return value indicates the comparison result between `aKey` and the entry (similar to `strcmp()`), i.e.,
+     * zero means perfect match, positive (> 0) indicates `aKey` is larger than entry, and negative indicates `aKey` is
+     * smaller than entry.
```

Comment:
```suggestion
     * The return value indicates the comparison result between @p aKey and the entry (similar to `strcmp()`), i.e.,
     * zero means perfect match, positive (> 0) indicates @p aKey is larger than entry, and negative indicates @p aKey is
     * smaller than entry.
```

Reply:
I changed this.

One thing related to `@p`: I somehow always assumed that it indicates (in doxygen) that the next term is actually a parameter. I think we use it in OT more or less under this assumption.
 
I learned recently that `@p` may just be a formatting indicator and practically same as using backtick in doxygen.